clap_builder/src/builder/value_hint.rs

```diff
@@ -65,6 +65,35 @@ pub enum ValueHint {
     Url,
     /// Email address.
     EmailAddress,
+
+    /// Custom value hint. This is used to provide a hint that is not covered by the predefined
+    Custom,
+}
+
+/// This trait is used to provide users a way to add custom value hint to the argument.
+/// This is useful when predefined value hints are not enough.
+pub trait CustomValueHint: core::fmt::Debug + Send + Sync + CustomValueHintClone {
```

Comment:
After #5618, this no longer needs to be in `clap_builder`.

`clap_complete`s `unstable-dynamic` will need to enable `clap`s `unstable-ext` feature

We'll need a struct that impl's `ArgExt`
```rust
pub struct ArgCompleter(Arc<dyn CustomCompleter>);

impl ArgCompleter {
    pub fn new<C: CustomCompleter>(completer: C) -> Self {
        Self(Arc::new(completer))
    }
}
```

```rust
// user code
arg.add(ArgCompleter::new(|| Some(["1", "2", "3"]....)
...
// clap_complete code
arg.get::<ArgCompleter>();
```